renderdoc/android/android_patch.cpp

```diff
@@ -520,6 +536,107 @@ bool IsDebuggable(const std::string &deviceID, const std::string &packageName)
 
   return pkgFlags.find("DEBUGGABLE") != std::string::npos;
 }
+
+rdcarray<rdcstr> GetActivitiesForPackage(const std::string &deviceID, const std::string &package)
+{
+  std::string defaultActivity = Android::GetDefaultActivityForPackage(deviceID, package);
+  rdcarray<rdcstr> activities;
+  activities.push_back(defaultActivity);
+
+  std::string aapt = getToolPath(ToolDir::BuildTools, "aapt", false);
+  std::string pkgPath = Android::GetPathForPackage(deviceID, package) + "base.apk";
+  std::string tmpDir = FileIO::GetTempFolderFilename();
+  std::string origAPK(tmpDir + package + ".orig.apk");
+
+  if(!Android::TryToPullApk(deviceID, pkgPath, origAPK, package))
```

Comment:
Pulling the APK from the device for any normal UI interaction is not acceptable as this is extremely slow due to how terrible adb is.

A quick google search suggests that it's possible to enumerate activities for all packages with `adb shell dumpsys package` without pulling any APKs and parsing the output. Why is this implemented using slow APK pulling & dumping instead?

Reply:
`adb shell dumpsys package` only provides the Activity which has the intent filter `android.intent.action.MAIN`. It doesn't list every Activity of the application. `adb pull` run on a background thread and called only once per application.

Reply:
Looking at my `dumpsys` output it definitely includes more activities than that, my guess is maybe it lists activities with any intent at all but I'm not sure what logic it uses. It seems like that's the list of activities that are intended to be public-facing anyway, so it's reasonable to say that only activities that are marked as public or whatever is required to show up in `dumpsys` can be browsed. As I understand it, it would be very minimal effort for a user to add an intent to their activity if the wanted to debug it through RenderDoc.

Using `adb pull` is unacceptable regardless of whether it's on a background thread or not. Either there's a solution that only uses adb to (relatively) quickly query, or there is no solution at all.